pkg/controller/statefulset/stateful_set_utils_test.go

```diff
@@ -417,71 +616,623 @@ func TestUpdateClaimOwnerRefForSetAndPod(t *testing.T) {
+				},
+				{
+					APIVersion: "someone",
+					Kind:       "Else",
+					Name:       "foo",
+				},
+			},
+		},
+		{
+			name: "missing controller",
+			originalRefs: []metav1.OwnerReference{
+				{
+					APIVersion: "apps/v1",
+					Kind:       "StatefulSet",
+					Name:       "sts",
+					UID:        "123",
+				},
+			},
+			expectedRefs: []metav1.OwnerReference{
+				{
+					APIVersion: "apps/v1",
+					Kind:       "StatefulSet",
+					Name:       "sts",
+					UID:        "123",
+					Controller: ptr.To(true),
+				},
+			},
+		},
+		{
+			name: "matching name but missing",
+			originalRefs: []metav1.OwnerReference{
+				{
+					APIVersion: "someone",
+					Kind:       "else",
+					Name:       "sts",
+					UID:        "456",
+				},
+			},
+			expectedRefs: []metav1.OwnerReference{
+				{
+					APIVersion: "someone",
+					Kind:       "else",
+					Name:       "sts",
+					UID:        "456",
+				},
+				{
+					APIVersion: "apps/v1",
+					Kind:       "StatefulSet",
+					Name:       "sts",
+					UID:        "123",
+					Controller: ptr.To(true),
+				},
+			},
+		},
+		{
+			name:         "not present",
+			originalRefs: []metav1.OwnerReference{},
+			expectedRefs: []metav1.OwnerReference{
+				{
+					APIVersion: "apps/v1",
+					Kind:       "StatefulSet",
+					Name:       "sts",
+					UID:        "123",
+					Controller: ptr.To(true),
+				},
+			},
+		},
+		{
+			name: "controller, but no UID",
+			originalRefs: []metav1.OwnerReference{
+				{
+					APIVersion: "apps/v1",
+					Kind:       "StatefulSet",
+					Name:       "sts",
+					Controller: ptr.To(true),
+				},
+			},
+			expectedRefs: []metav1.OwnerReference{
+				{
+					APIVersion: "apps/v1",
+					Kind:       "StatefulSet",
+					Name:       "sts",
+					UID:        "123",
+					Controller: ptr.To(true),
+				},
+			},
+		},
+		{
+			name: "neither controller nor UID",
+			originalRefs: []metav1.OwnerReference{
+				{
+					APIVersion: "apps/v1",
+					Kind:       "StatefulSet",
+					Name:       "sts",
+				},
+			},
+			expectedRefs: []metav1.OwnerReference{
+				{
+					APIVersion: "apps/v1",
+					Kind:       "StatefulSet",
+					Name:       "sts",
+					UID:        "123",
+					Controller: ptr.To(true),
+				},
+			},
+		},
+	}
+	for _, tc := range testCases {
+		_, ctx := ktesting.NewTestContext(t)
+		logger := klog.FromContext(ctx)
+		set := apps.StatefulSet{}
+		set.Name = "sts"
+		set.Spec.Replicas = ptr.To(int32(1))
+		set.SetUID("123")
+		set.Spec.PersistentVolumeClaimRetentionPolicy = &apps.StatefulSetPersistentVolumeClaimRetentionPolicy{
+			WhenScaled:  apps.DeletePersistentVolumeClaimRetentionPolicyType,
+			WhenDeleted: apps.DeletePersistentVolumeClaimRetentionPolicyType,
+		}
+		pod := v1.Pod{}
+		pod.Name = "pod-0"
+		pod.SetUID("456")
+		claim := v1.PersistentVolumeClaim{}
+		claim.SetOwnerReferences(tc.originalRefs)
+		updateClaimOwnerRefForSetAndPod(logger, &claim, &set, &pod)
+		if ownerRefsChanged(tc.expectedRefs, claim.GetOwnerReferences()) {
```

Comment:
can we simplify this and just do `DeepEqual`?

Reply:
DeepEqual won't work if the order of the references changes?

Being order-independent will make the test less fragile and less of a change-detector.

Reply:
Thanks